compiler/rustc_lint/src/types/improper_ctypes.rs

```diff
@@ -429,6 +619,111 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         }
     }
 
+    fn visit_struct_or_union(
+        &mut self,
+        state: VisitorState,
+        ty: Ty<'tcx>,
+        def: AdtDef<'tcx>,
+        args: GenericArgsRef<'tcx>,
+    ) -> FfiResult<'tcx> {
+        debug_assert!(matches!(def.adt_kind(), AdtKind::Struct | AdtKind::Union));
+        use FfiResult::*;
+
+        if !def.repr().c() && !def.repr().transparent() {
+            return FfiUnsafe {
+                ty,
+                reason: if def.is_struct() {
+                    fluent::lint_improper_ctypes_struct_layout_reason
+                } else {
+                    fluent::lint_improper_ctypes_union_layout_reason
+                },
+                help: if def.is_struct() {
+                    Some(fluent::lint_improper_ctypes_struct_layout_help)
+                } else {
+                    // FIXME(ctypes): confirm that this makes sense for unions once #60405 / RFC2645 stabilises
```

Comment:
```suggestion
                    // FIXME(#60405): confirm that this makes sense for unions once #60405 / RFC2645 stabilises
```

Since this will be fixed by something other than the ctypes refactoring